ansible_base/authentication/utils/claims.py

```diff
@@ -268,8 +277,104 @@ def update_user_claims(user: Optional[AbstractUser], database_authenticator: Aut
     return user
 
 
+def process_organization_and_team_memberships(results) -> None:
+    # Extract organizations where the user is a member
+    org_list = set()
+    for org_name, is_member in results['claims']['organization_membership'].items():
+        if is_member:
+            org_list.add(org_name)
+
+    # Build a mapping of teams to their respective organizations, filtering out non-members
+    team_map = {}
+    for org_name, teams in results['claims']['team_membership'].items():
+        for team_name, is_member in teams.items():
+            if is_member:
+                team_map[team_name] = org_name
+
+    # Create organizations and teams based on the membership data
+    create_orgs_and_teams(org_list, team_map)
+
+
+def create_orgs_and_teams(org_list: set[str], team_map: dict[str, str]) -> None:
+    # Ensure unique organization names and gather all team names
+    all_orgs = set(org_list) | set(team_map.values())
+    all_teams = list(team_map.keys())
+
+    # Load existing organizations
+    existing_orgs = load_existing_orgs(all_orgs)
+
+    # Create missing organizations
+    create_missing_orgs(all_orgs, existing_orgs)
+
+    # Load existing teams
+    existing_teams = load_existing_teams(all_teams)
+
+    # Create missing teams
+    create_missing_teams(all_teams, team_map, existing_orgs, existing_teams)
+
+
+def load_existing_orgs(org_names: set[str]) -> dict[str, int]:
+    existing_orgs = {org.name: org.id for org in Organization.objects.filter(name__in=org_names)}
+    return existing_orgs
+
+
+def create_missing_orgs(org_names: set[str], existing_orgs: dict[str, int]) -> None:
+    for org_name in org_names:
+        if org_name not in existing_orgs:
+            logger.info(f"creating org {org_name}")
+            new_org, _ = Organization.objects.get_or_create(name=org_name)
+            existing_orgs[org_name] = new_org.id
+
+
+def load_existing_teams(team_names: Iterable[str]) -> set[str]:
+    existing_teams = set(Team.objects.filter(name__in=team_names).values_list('name', flat=True))
+    return existing_teams
+
+
+def create_missing_teams(team_names: Iterable[str], team_map: dict[str, str], existing_orgs: dict[str, int], existing_teams: set[str]) -> None:
+    for team_name in team_names:
+        if team_name not in existing_teams:
```

Comment:
Team names are not unique. What happens if I have a team named "My team" in "Org A" but the login wants a "My Team" in "Org B". This would detect "My Team" from "Org A" but not create the team in "Org B" or am I missing something?

Reply:
That is currently unaccounted for. I'm fixing it ..

Reply:
resolved with latest commit